### PR TITLE
Use SELF for Builder Self Type Parameter (part2)

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
See #5684 for the motivation.

Modifications:
- Refactored builder classes to use the SELF type parameter for the builder's self-referencing methods.

Result:
- Reduced the need for cumbersome and error-prone method overrides.
- (Breaking) These are all annotated with `@UnstableApi`.
  - `AbstractDnsResolverBuilder` is now `AbstractDnsResolverBuilder<SELF>`
  - `AbstractRuleBuilder` is now `AbstractRuleBuilder<SELF>`
  - `AbstractRuleWithContentBuilder` is now `AbstractRuleWithContentBuilder<SELF>`
  - `AbstractCircuitBreakerMappingBuilder` is now `AbstractCircuitBreakerMappingBuilder<SELF>`
  - `AbstractDynamicEndpointGroupBuilder` is now `AbstractDynamicEndpointGroupBuilder<SELF>`
  - `DynamicEndpointGroupSetters` is now `DynamicEndpointGroupSetters<SELF>`
  - `AbstractCuratorFrameworkBuilder` is now `AbstractCuratorFrameworkBuilder<SELF>`

### CODE REVIEW

**Review by jrhee17 — APPROVED** [review]
Agree with the overall direction 👍  I'd rather not go through each change since I think it's more efficient that we fix them as we go along. Thanks @minwoox 🙇 👍 🙇

**Review by ikhoon — APPROVED** [review]

File: core/src/main/java/com/linecorp/armeria/client/AbstractDnsResolverBuilder.java [hunk 1]
```diff
@@ -92,6 +92,14 @@ public abstract class AbstractDnsResolverBuilder {
      */
     protected AbstractDnsResolverBuilder() {}
 
+    /**
+     * Return this.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/AbstractDnsResolverBuilder.java:96]: Global comment:
```suggestion
     * Returns {@code this}.
```

File: core/src/main/java/com/linecorp/armeria/client/AbstractRuleWithContentBuilder.java [hunk 2]
```diff
@@ -36,7 +36,9 @@
  * @param <T> the response type
  */
 @UnstableApi
-public abstract class AbstractRuleWithContentBuilder<T extends Response> extends AbstractRuleBuilder {
+public abstract class AbstractRuleWithContentBuilder
+        <T extends Response, SELF extends AbstractRuleWithContentBuilder<T, SELF>>
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/AbstractRuleWithContentBuilder.java:40]: Global comment: Maybe `SELF` should come first?

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/AbstractRuleWithContentBuilder.java]: Thanks! Updated. 👍

**Review by trustin — APPROVED** [review]
Thanks, @minwoox!

**Pull request merged**